### ISSUE TITLE
Derived protocol with `Self` used in base protocol still works incorrectly.

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#5087]
**Describe the bug**
Related to #4966. Seems to still work incorrectly over some cases.

**To Reproduce**
See sample code below.

**Expected behavior**
`MutableSetLike1` and `MutableSetLike2` should work equally as protocols.

**Screenshots or Code**
```python
from __future__ import annotations

from typing import Protocol, TypeVar
from typing_extensions import Self

_V_t = TypeVar("_V_t")
_V_co = TypeVar("_V_co")

class SetLike(Protocol[_V_co]):
    def __and__(self, right: Self, /) -> SetLike[_V_co]: ...
    # Other methods are omitted.

class MutableSetLike1(SetLike[_V_t], Protocol[_V_t]):
    def add(self, o: _V_t, /) -> None: ...
    # Other methods are omitted.
    
class MutableSetLike2(Protocol[_V_t]):
    def __and__(self, right: Self, /) -> SetLike[_V_t]: ...
    def add(self, o: _V_t, /) -> None: ...
    # Other methods are omitted.

SOME_SET: set[str] = {"A", "B", "C"}

a: SetLike[str] = SOME_SET
b: MutableSetLike1[str] = SOME_SET
c: MutableSetLike2[str] = SOME_SET
```
![image](https://github.com/microsoft/pyright/assets/15389744/67efc34a-27e0-4cb2-8ea9-ced12c9766a7)

**VS Code extension or command-line**

Pylance v2023.5.11 (Pyright 1.1.306).

**Additional context**

Mypy playground: https://mypy-play.net/?mypy=latest&python=3.10&gist=d68e3371c260e7b333d059194b1b9774

**Comment by erictraut** [comment]
Thanks for the bug report. This was a different (but somewhat related) issue than the issue I fixed in #4966. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.308, which I just published. It will also be included in a future release of pylance.
